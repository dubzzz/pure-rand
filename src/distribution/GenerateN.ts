import type { RandomGenerator } from '../types/RandomGenerator';
import { unsafeGenerateN } from './UnsafeGenerateN';

export function generateN(rng: RandomGenerator, num: number): [number[], RandomGenerator] {
  const nextRng = rng.clone();
  const out = unsafeGenerateN(nextRng, num);
  return [out, nextRng];
}
