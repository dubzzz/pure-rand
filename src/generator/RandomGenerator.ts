export interface RandomGenerator {
  /** Produce a fully independent clone of the current instance */
  clone(): RandomGenerator;
  /**
   * Generate next random value along with the next generator (does not impact current instance).
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  next(): [number, RandomGenerator];
  /** Compute the jumped generator (does not impact current instance) */
  jump?(): RandomGenerator;
  /**
   * Generate next value BUT alters current generator.
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  unsafeNext(): number;
  /** Jump current generator */
  unsafeJump?(): void;
}

export function unsafeGenerateN(rng: RandomGenerator, num: number): number[] {
  const out: number[] = [];
  for (let idx = 0; idx != num; ++idx) {
    out.push(rng.unsafeNext());
  }
  return out;
}

export function generateN(rng: RandomGenerator, num: number): [number[], RandomGenerator] {
  const nextRng = rng.clone();
  const out = unsafeGenerateN(nextRng, num);
  return [out, nextRng];
}

export function unsafeSkipN(rng: RandomGenerator, num: number): void {
  for (let idx = 0; idx != num; ++idx) {
    rng.unsafeNext();
  }
}

export function skipN(rng: RandomGenerator, num: number): RandomGenerator {
  const nextRng = rng.clone();
  unsafeSkipN(nextRng, num);
  return nextRng;
}
