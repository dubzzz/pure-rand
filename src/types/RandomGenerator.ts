export interface RandomGenerator {
  /** Produce a fully independent clone of the current instance */
  clone(): RandomGenerator;
  /**
   * Generate next random value along with the next generator (does not impact current instance).
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  next(): [number, RandomGenerator];
  /** Compute the jumped generator (does not impact current instance) */
  jump?(): RandomGenerator;
  /**
   * Generate next value BUT alters current generator.
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  unsafeNext(): number;
  /** Jump current generator */
  unsafeJump?(): void;
  /** Access to the internal state of a RandomGenerator in a read-only fashion */
  getState?(): readonly number[];
}
