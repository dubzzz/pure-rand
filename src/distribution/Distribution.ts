import { RandomGenerator } from '../generator/RandomGenerator';

/**
 * Generate random value based on a given RandomGenerator.
 * Return the generated value and an offsetted version of the RandomGenerator (never alters the source rng).
 * @public
 */
export type Distribution<T> = (rng: RandomGenerator) => [T, RandomGenerator];
