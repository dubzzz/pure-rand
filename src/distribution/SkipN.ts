import { RandomGenerator } from '../types/RandomGenerator';
import { unsafeSkipN } from './UnsafeSkipN';

export function skipN(rng: RandomGenerator, num: number): RandomGenerator {
  const nextRng = rng.clone();
  unsafeSkipN(nextRng, num);
  return nextRng;
}
