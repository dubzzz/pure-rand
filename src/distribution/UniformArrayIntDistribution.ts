import { Distribution } from './Distribution';
import { RandomGenerator } from '../generator/RandomGenerator';
import { ArrayInt } from './internals/ArrayInt';
import { unsafeUniformArrayIntDistribution } from './UnsafeUniformArrayIntDistribution';

/**
 * Uniformly generate random ArrayInt values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 *
 * @public
 */
function uniformArrayIntDistribution(from: ArrayInt, to: ArrayInt): Distribution<ArrayInt>;
/**
 * Uniformly generate random ArrayInt values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 * @param rng - Instance of RandomGenerator to extract random values from
 *
 * @public
 */
function uniformArrayIntDistribution(from: ArrayInt, to: ArrayInt, rng: RandomGenerator): [ArrayInt, RandomGenerator];
function uniformArrayIntDistribution(from: ArrayInt, to: ArrayInt, rng?: RandomGenerator) {
  if (rng != null) {
    const nextRng = rng.clone();
    return [unsafeUniformArrayIntDistribution(from, to, nextRng), nextRng];
  }
  return function (rng: RandomGenerator) {
    const nextRng = rng.clone();
    return [unsafeUniformArrayIntDistribution(from, to, nextRng), nextRng];
  };
}

export { uniformArrayIntDistribution };
