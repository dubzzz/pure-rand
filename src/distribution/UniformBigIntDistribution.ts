import { Distribution } from './Distribution';
import { RandomGenerator } from '../generator/RandomGenerator';
import { unsafeUniformBigIntDistribution } from './UnsafeUniformBigIntDistribution';

/**
 * Uniformly generate random bigint values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 *
 * @public
 */
function uniformBigIntDistribution(from: bigint, to: bigint): Distribution<bigint>;
/**
 * Uniformly generate random bigint values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 * @param rng - Instance of RandomGenerator to extract random values from
 *
 * @public
 */
function uniformBigIntDistribution(from: bigint, to: bigint, rng: RandomGenerator): [bigint, RandomGenerator];
function uniformBigIntDistribution(from: bigint, to: bigint, rng?: RandomGenerator) {
  if (rng != null) {
    const nextRng = rng.clone();
    return [unsafeUniformBigIntDistribution(from, to, nextRng), nextRng];
  }
  return function (rng: RandomGenerator) {
    const nextRng = rng.clone();
    return [unsafeUniformBigIntDistribution(from, to, nextRng), nextRng];
  };
}

export { uniformBigIntDistribution };
