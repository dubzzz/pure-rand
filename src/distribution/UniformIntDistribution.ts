import { Distribution } from './Distribution';
import { RandomGenerator } from '../generator/RandomGenerator';
import { unsafeUniformIntDistribution } from './UnsafeUniformIntDistribution';

/**
 * Uniformly generate random integer values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 *
 * @public
 */
function uniformIntDistribution(from: number, to: number): Distribution<number>;
/**
 * Uniformly generate random integer values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 * @param rng - Instance of RandomGenerator to extract random values from
 *
 * @public
 */
function uniformIntDistribution(from: number, to: number, rng: RandomGenerator): [number, RandomGenerator];
function uniformIntDistribution(from: number, to: number, rng?: RandomGenerator) {
  if (rng != null) {
    const nextRng = rng.clone();
    return [unsafeUniformIntDistribution(from, to, nextRng), nextRng];
  }
  return function (rng: RandomGenerator) {
    const nextRng = rng.clone();
    return [unsafeUniformIntDistribution(from, to, nextRng), nextRng];
  };
}

export { uniformIntDistribution };
