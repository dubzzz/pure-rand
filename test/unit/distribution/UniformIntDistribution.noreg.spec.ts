import fc from 'fast-check';

import { uniformIntDistribution } from '../../../src/distribution/UniformIntDistribution';
import mersenne from '../../../src/generator/MersenneTwister';
import { RandomGenerator } from '../../../src/generator/RandomGenerator';

describe('uniformIntDistribution [non regression]', () => {
  it.each`
    from                       | to                         | topic
    ${0}                       | ${2 ** 3 - 1}              | ${"range of size divisor of mersenne's one"}
    ${0}                       | ${2 ** 3 - 2}              | ${"range of size divisor of mersenne's one minus one"}
    ${0}                       | ${2 ** 3}                  | ${"range of size divisor of mersenne's one plus one"}
    ${48}                      | ${69}                      | ${'random range'}
    ${-0x80000000}             | ${0x7fffffff}              | ${"mersenne's range"}
    ${-0x80000000}             | ${0x7fffffff - 1}          | ${"mersenne's range minus one"}
    ${-0x80000000}             | ${0x7fffffff + 1}          | ${"mersenne's range plus one"}
    ${0}                       | ${2 ** 40 - 1}             | ${"range of size multiple of mersenne's one"}
    ${0}                       | ${2 ** 40 - 2}             | ${"range of size multiple of mersenne's one minus one"}
    ${0}                       | ${2 ** 40}                 | ${"range of size multiple of mersenne's one plus one"}
    ${Number.MIN_SAFE_INTEGER} | ${Number.MAX_SAFE_INTEGER} | ${'full integer range'}
  `('Should not change its output in range ($from, $to) except for major bumps', ({ from, to }) => {
    // Remark:
    // ========================
    // This test is purely there to ensure that we do not introduce any regression
    // during a commit without noticing it.
    // The values we expect in the output are just a snapshot taken at a certain time
    // in the past. They might be wrong values with bugs.

    let rng: RandomGenerator = mersenne(0);
    const distribution = uniformIntDistribution(from, to);

    const values: number[] = [];
    for (let idx = 0; idx !== 10; ++idx) {
      const [v, nrng] = distribution(rng);
      values.push(v);
      rng = nrng;
    }
    expect(values).toMatchSnapshot();
  });

  it('Should always generate values within the range [from ; to]', () =>
    fc.assert(
      fc.property(fc.integer().noShrink(), fc.maxSafeInteger(), fc.maxSafeInteger(), (seed, a, b) => {
        const [from, to] = a < b ? [a, b] : [b, a];
        const [v, _nrng] = uniformIntDistribution(from, to)(mersenne(seed));
        return v >= from && v <= to;
      })
    ));
});
