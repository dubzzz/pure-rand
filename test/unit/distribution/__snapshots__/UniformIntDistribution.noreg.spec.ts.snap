// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uniformIntDistribution [non regression] Should not change its output in range (-2147483648, 2147483646) except for major bumps 1`] = `
[
  -1937831252,
  -1748719057,
  -1223252363,
  -668873536,
  -1706118333,
  -610118917,
  -1954711869,
  -656048793,
  1819583497,
  -1616781613,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (-2147483648, 2147483647) except for major bumps 1`] = `
[
  -1937831252,
  -1748719057,
  -1223252363,
  -668873536,
  -1706118333,
  -610118917,
  -1954711869,
  -656048793,
  1819583497,
  -1616781613,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (-2147483648, 2147483648) except for major bumps 1`] = `
[
  -1748719057,
  1277901399,
  243580376,
  1171049868,
  2051556033,
  -806729177,
  1686997841,
  602801999,
  2034131043,
  -855081807,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (-9007199254740991, 9007199254740991) except for major bumps 1`] = `
[
  8737458527447601,
  -2631726847111999,
  -4869447699246340,
  8293355706744169,
  5809858626576085,
  -1881170402421005,
  -7398455699097000,
  -5844700774482471,
  5236255647124877,
  2596200260321985,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (0, 6) except for major bumps 1`] = `
[
  2,
  1,
  5,
  0,
  6,
  0,
  4,
  1,
  6,
  3,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (0, 7) except for major bumps 1`] = `
[
  4,
  7,
  5,
  0,
  3,
  3,
  3,
  7,
  1,
  3,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (0, 8) except for major bumps 1`] = `
[
  6,
  7,
  0,
  4,
  5,
  4,
  5,
  8,
  3,
  7,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (0, 1099511627774) except for major bumps 1`] = `
[
  739133139503,
  503989783744,
  289300173563,
  839010057575,
  39185407699,
  93992703730,
  158044207703,
  303038774744,
  381275655564,
  253307142849,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (0, 1099511627775) except for major bumps 1`] = `
[
  86298110511,
  753097886912,
  847645922043,
  821830188391,
  1027027885779,
  1073245247218,
  570361068119,
  1041773149656,
  1081355324812,
  317731652289,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (0, 1099511627776) except for major bumps 1`] = `
[
  86298110511,
  753097886912,
  847645922043,
  821830188391,
  1027027885779,
  1073245247218,
  570361068119,
  1041773149656,
  1081355324812,
  317731652289,
]
`;

exports[`uniformIntDistribution [non regression] Should not change its output in range (48, 69) except for major bumps 1`] = `
[
  56,
  53,
  69,
  60,
  65,
  49,
  61,
  57,
  67,
  65,
]
`;
